flang/lib/Optimizer/Transforms/ArrayValueCopy.cpp

```diff
@@ -101,80 +101,136 @@ class ArrayCopyAnalysis {
 };
 } // namespace
 
-// Recursively trace operands to find all array operations relating to the
-// values merged.
-static void populateSets(llvm::SmallVectorImpl<mlir::Operation *> &reach,
-                         llvm::SmallPtrSetImpl<mlir::Value> &visited,
-                         mlir::Value val) {
-  if (!val || visited.contains(val))
-    return;
-  visited.insert(val);
+namespace {
 
-  if (auto *op = val.getDefiningOp()) {
-    // `val` is defined by an Op, process the defining Op.
-    // If `val` is defined by a region containing Op, we want to drill down and
-    // through that Op's region(s).
-    LLVM_DEBUG(llvm::dbgs() << "popset: " << val << '\n');
+/// Helper class to collect all array operations that produced an array value.
+class ReachCollector {
+public:
+  ReachCollector(llvm::SmallVectorImpl<mlir::Operation *> &reach,
+                 mlir::Region *loopRegion)
+      : reach{reach}, loopRegion{loopRegion} {}
+
+  // Recursively trace operands to find all array operations relating to the
+  // values merged.
+  void collectArrayAccessFrom(mlir::Value val) {
+    if (!val || visited.contains(val))
+      return;
+    visited.insert(val);
+
+    if (auto *op = val.getDefiningOp()) {
+      // `val` is defined by an Op, process the defining Op.
+      // If `val` is defined by a region containing Op, we want to drill down
+      // and through that Op's region(s).
+      LLVM_DEBUG(llvm::dbgs() << "popset: " << val << '\n');
+      auto popFn = [&](auto rop) {
+        auto resNum = val.cast<mlir::OpResult>().getResultNumber();
+        llvm::SmallVector<mlir::Value> results;
+        rop.resultToSourceOps(results, resNum);
+        for (auto u : results)
+          collectArrayAccessFrom(u);
+      };
+      if (auto rop = mlir::dyn_cast<DoLoopOp>(op)) {
+        popFn(rop);
+        return;
+      }
+      if (auto rop = mlir::dyn_cast<IterWhileOp>(op)) {
+        popFn(rop);
+        return;
+      }
+      if (auto rop = mlir::dyn_cast<fir::IfOp>(op)) {
+        popFn(rop);
+        return;
+      }
+
+      if (mlir::isa<fir::AllocaOp, fir::AllocMemOp>(op))
```

Comment:
I'm not totally clear as to why the look at these dynamic allocations was added here. The relevant scope should be from the dominating `array_loads` over the region to the post-dominating `array_merge_store`. Some questions arise: what about arrays that are not locally allocated? Does this discriminate scalars somehow? What happens if we reach an alloca of an array, but it is used in more than one copy-in/copy-out context?